### ISSUE TITLE
[Discussion] Ways to improve multi-thread handling beyond the current possibilites

### ISSUE THREAD

**Issue by Jojo-1000** [robinrodricks/FluentFTP.GnuTLS#88]
Trying to use multiple `AsyncFtpClient` instances results in an access violation:

```
System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at FluentFTP.GnuTLS.GnuTlsInternalStream.Dispose()
   at FluentFTP.GnuTLS.GnuTlsStream.Dispose()
   at FluentFTP.FtpSocketStream.Dispose(Boolean disposing)
   at System.IO.Stream.Close()
   at FluentFTP.Client.BaseClient.BaseFtpClient.Dispose()
   at FluentFTP.Client.BaseClient.BaseFtpClient.Finalize()
```

## Test snippet
```c#
private static async Task TestFtp()
{
    FluentFTP.AsyncFtpClient client = new FluentFTP.AsyncFtpClient("127.0.0.1", 21, new FluentFTP.FtpConfig()
    {
        DataConnectionType = FluentFTP.FtpDataConnectionType.AutoPassive,
        EncryptionMode = FluentFTP.FtpEncryptionMode.Explicit,
        SslProtocols = System.Security.Authentication.SslProtocols.Tls12,
        CustomStream = typeof(GnuTlsStream),
        CustomStreamConfig = new GnuConfig()
        {
            SecuritySuite = FluentFTP.GnuTLS.Enums.GnuSuite.Secure128,
            SetALPNControlConnection = string.Empty,
            SetALPNDataConnection = string.Empty
        }

    })
    {
        Credentials = new System.Net.NetworkCredential("user", "asdf")
    };
    client.ValidateCertificate += (BaseFtpClient control, FtpSslValidationEventArgs e) => { e.Accept = true; };
    var listing = await client.GetListing("/temp");
    foreach (var item in listing)
    {
        Console.WriteLine(item.FullName);
    }
}
public static int Main(string[] args)
{
    var t1 = Task.Run(TestFtp);
    var t2 = Task.Run(TestFtp);
    t1.Wait();
    t2.Wait();
    return 0;
}
```
Connecting to FileZilla Server version 1.7.2.

If I comment out t2, there is no exception. Both list tasks print all of the files before crashing, it seems like a resource is shared and deleted twice.

**Comment by FanDjango** [comment]
Yeah, there's a config that you can set to tell him: don't dispose for when you need thread safety. Go through the recent closed issues, you'll find it.

**Comment by Jojo-1000** [comment]
https://github.com/robinrodricks/FluentFTP/issues/1273

Do I understand correctly that there will only be one parallel upload/download at a time?

**Comment by FanDjango** [comment]
No. Potentially thousands in parallel

**Comment by Jojo-1000** [comment]
That is great to hear.

In our application (backup software) there can be a long time between connections. It would be nice to close the gnutls library to use fewer resources.

I think your init/deinit problem can be solved by reference counting the number of open users. Then the option would no longer be necessary and there is no open library.

If that sounds good, I can try and make a pull request.

**Comment by FanDjango** [comment]
@Jojo-1000 

The FluentFPT.GnuTLS "extension" was prompted by a flurry of problems that surfaced last year or even earlier. No need to reiterate them here.

I completed most of the coding (we are not yet at 100%, by far) in the first quarter of this year. Even with some features missing it was deemed a good idea to release this and work onwards depending on feedback that is coming in.

There has not been much such feedback, so there is no real indication how many users are actually profiting from this. As usual, you only hear from the community out there when something goes wrong.

The thread-safety issue for parallel use has come up only once previously, you are the second such reporter. By the way, it was not surprising, the limitations of the initial code were clear.

So: right now, we are at the stage that we hope for more feedback on "**it now works in a parallel-usage environment if you set this parameter**", and to enable that, we have initially created that rather strange new config parameter.

In the end, your suggestion of a thread-safe "in-use" counter **is (probably) the best way** to go. If you want to open a PR for that, **I would be enthusiastic to see that happening**.

Concerning the resource use of "keeping the lib initialiazed", there is actually not a huge footprint to worry about using the current solution - just like I would worry about repeated de-inits followed by re-inits with a usage count solution. Maybe one would in the end offer multiple possible solutions for this problem depending on the usage szenario of the user.

**Comment by Jojo-1000** [comment]
> just like I would worry about repeated de-inits followed by re-inits with a usage count solution.

Maybe instead of a direct use count, it could be run in the destructor/finalizer of an object. That way the library would be closed when the garbage collector runs, so I think there is some delay between using and closing it. But I am not sure how C# garbage collection works in detail.

**Comment by FanDjango** [comment]
I understand your idea. 

Hmmm.

First thought would be: Let's offer to the user the following options:

1. Single thread operation. Init and De-Init bracket all uses of the **single** `GnuTlsStream`

It has the advantage of being linear in operation and good for debugging a single usage of the stream.

2. Multi-thread operation with no De-Init

Cleanest way to ensure trouble free thread safety.

3. (Future) Multi-thread operation by way of a "use-count"

Less resource usage, (possibly) more overhead

4. (Future) Externalize Init and De-Init in some way for the user to execute as and when he sees fit.

That's actually the way the base GnuTLS library does it: By giving you full control over when and how you Init and De-Init the lib.

**Comment by fredericDelaporte** [comment]
3. would be a regression for my use case. It is a background service polling a FTP server every minute. With de-init enabled, some kind of resource leak seems to occurs (handles count increasing) and the service is abruptly killed up to two times a day. (No exception logged, just an abnormal service stop log in the Windows event system log.)

(For completeness, this service also do some uploads, most guaranteed to not occur during polling, but a few others may occur in another thread during polling. This service also uses the async FTP client.)

I think 4. should be done, maybe through some static class/method, or an initializer/deinitializer class to instantiate and run once. 

And then do some breaking change to avoid surprising pitfalls in multi-threaded context: no more on-the-fly init if not explicitly initialized, but an InvalidOperationException stating the lib needs an explicit initialization. Eventually a less breaking change could be done instead: switch deinit to false by default.

Being exposed to thread safety troubles by default is a flaw in my opinion. The library should not stay in that situation.

**Comment by Jojo-1000** [comment]
From the [docs of global_init](https://www.gnutls.org/manual/html_node/Core-TLS-API.html#gnutls_005fglobal_005finit), it is not necessary to call it at all since GnuTLS 3.3. It also states that there is internal reference counting for init/deinit and it is thread safe. So to fix the crash I think the only change would need to be removing the if intialized checks in the GnuTLS stream and always call init/deinit. Then you can still use the option to keep the library initialized if you prefer, but it would hopefully no longer crash by default.

**Comment by FanDjango** [comment]
Thanks for the comments/inputs.

Is there anything keeping you from testing/implementing these ideas?

In the next days, I will play with this also. It is correct that from GnuTLS 3.3 onwards, a repeated Init call is "ignored". And that de-inits are supposed to be handled in a graceful fashion.

The problem lies in the "cred.dispose", possibly? I will look at it in more detail.

**Comment by FanDjango** [comment]
> Being exposed to thread safety troubles by default is a flaw in my opinion. The library should not stay in that situation

I agree. I invested a LOT of time to get it to where it is now (from zero) and I appreciate any help in this direction that may be available. My personal use cases **do not** extend to multi-threaded operations.

**Comment by Jojo-1000** [comment]
Unfortunately it seems that my changes are not enough to fix the crash. There also seems to be multiple points where the crash occurs, sometimes in a `Read()` with an expired session pointer, and sometimes in `Credentials.Dispose` with an expired credentials pointer. Since I have the build set up now, I hope I will be able to fix this in the next days.

**Comment by FanDjango** [comment]
Ok, so now lets play with this and stress test it. Nice changes - splendid that we can get on with this topic. Thanks!

**Comment by FanDjango** [comment]
I've decided to utilize the "use-count" logic **all the time**. No need for the config variable DeInitGnuTls anmore. Those of you who have been using it, will get a compile error and can just remove it.

I know I stated this earlier:

> First thought would be: Let's offer to the user the following options:
Single thread operation. Init and De-Init bracket all uses of the single GnuTlsStream
It has the advantage of being linear in operation and good for debugging a single usage of the stream.
Multi-thread operation with no De-Init
Cleanest way to ensure trouble free thread safety.
(Future) Multi-thread operation by way of a "use-count"
Less resource usage, (possibly) more overhead
(Future) Externalize Init and De-Init in some way for the user to execute as and when he sees fit.

But this has the least complexity...

**Comment by fredericDelaporte** [comment]
Thanks for your effort on this topic, but as I told, the use-count logic without the de-init option for disabling it will likely cause a serious regression in my case.

So, I reformulate: currently, initializing / de-initializaing the GnuTLS library a lot of times causes some resources leaks ending up fatal-crashing the process (catch blocks not even executed). At least that is what happens in the case of my application. Deactivating the de-initialization has fixed these crashes for my application.

**Comment by FanDjango** [comment]
@fredericDelaporte Ok, don't worry. I am already working on providing an option to "continuously" keep the libs loaded.

It is just easier to code it "this way around".

The option **you** need will arrive soon.

**Comment by fredericDelaporte** [comment]
Ok, thanks.

**Comment by FanDjango** [comment]
@fredericDelaporte And also, you have not tested the current status, @Jojo-1000 had fixed a few loopholes in the way Dispose was coded. So you cannot be sure that the failure you described actually still happens.

**Comment by FanDjango** [comment]
@fredericDelaporte 

I did say:

> and if someone now wants, for some special reason, to disable possible unloading and re-loading*1), it will be easier to bring that in

**Comment by fredericDelaporte** [comment]
> Jojo-1000 had fixed a few loopholes in the way Dispose was coded

I have read that, yes, thus my "likely" wording. But checking if these loopholes were the reason for the fatal-crash cannot be done easily, since it involves running the application many days to ascertain it does not fatal-crash occasionally again. So, I had preferred to attempt to be on the safe side by voicing my preference for still being able to not have the library de-initialized each time all FTP clients are disposed of.

> I did say:

Seen too, but for me it was not clear whether it would be done anyway or only if someone ends-up reporting troubles due to not having that possibility.

**Comment by FanDjango** [comment]
@fredericDelaporte I understand. All is good. Working on it. I want it to be "elegant", better than the previous rather hurried solutions, now that we have a good handle on the logic.

**Comment by FanDjango** [comment]
So now we have the following status:

**Per default**, dll librarys are loaded on first use, and unloaded when stream disposed and use count is 0.

This behaviour, as default, works fine for single thread use and for multi-threaded concurrent use.

A user, who is observing too many library unloads/re-loads on account of his special use case, can disable unloading (and thus also the re-loading) of the libraries by using the special config variable `DllUnloadByUseCount` (default is **true**) and setting this to **false**.

Why might a user observe too many library unloads/re-loads and why might it be desirable to disable this?

1. Long running program performing ftp requests, constantly loading, transferring and then unloading until the next request comes in.
2. Single thread program transfering very many single files while connecting/disconnecting for each file.

a.) Worried about possible memory leaks
b.) Worried about performance penalty of unload/re-load process

**Comment by FanDjango** [comment]
```cs
		/// <summary>
		/// The FluentFTP.GnuTLS extension to FluentFTP will correctly handle multi-threaded
		/// concurrent operation by maintining a "use-count", just like the underlying GnuTLS
		/// library does. This use-count is checked on dispose of the GnuTlsStream and an unload
		/// of the GnuTls library .dlls only takes place when this value is zero.
		/// To disable unloading and to keep the libraries permanently loaded, set
		/// DllUnloadByUseCount to false.
		/// </summary>
		public bool DllUnloadByUseCount { get; set; } = true;
```

**Comment by FanDjango** [comment]
**released 1.0.21** with the above changes.

#### 1.0.21
  - Implement a use-count based load/unload logic for the .dll libraries (thanks @Jojo-1000)
  - Add logging for load and unload library actions
  - Add a config parm DllUnloadByUseCount to disable use-count based unloading
  - Remove the interim experimental parm DeInitGnuTLS that had a similar function
  - Improve cred and sess dispose handling (thanks @Jojo-1000)

We will see where this takes us. Please provide input as your schedules allow and thanks for the cooperation and input thus far.

**Comment by fredericDelaporte** [comment]
Hello,

I have tried 1.0.21 without setting `DllUnloadByUseCount` to false, leaving it being true: fatal crashes are back. After some ten hours of polling a FTP server every minute, disposing of the FTP client at each time, the service crashed again without any exception logged anywhere (neither our log nor in Windows events log), just a hard kill of the process. (System events log stating the service has unexpectedly stopped. No multi-threaded use of a FTP client was involved in this test.)

So, there is still an issue when repeatedly initializing/de-initializing GnuTLS. (Of course, I will go on deactivating this feature for production use. Or maybe I could change the code of that service to not release the FTP client but just close/re-open it.)

(Although this trouble is actually not multi-thread related, since we are discussing the init/deinit subject here, I have not opened another issue.)

**Comment by FanDjango** [comment]
> So, there is still an issue when repeatedly initializing/de-initializing GnuTLS. (Of course, I will go on deactivating this feature for production use. Or maybe I could change the code of that service to not release the FTP client but just close/re-open it.)

Interesting. Thanks for doing this. Since it has nothing to do with multithread operations, we can test this (probably, even I can test it) separately and also, we will do it in a separate issue than this one. This one is "Multi-Thread", so to speak. I will open one soon.

**Comment by FanDjango** [comment]
See #100
